### DISCUSSION TITLE
Offline - handle non-offline errors gracefully?

### DISCUSSION THREAD

**Discussion by robertherber** [urql-graphql/urql#2671] [Q&A]
When using the offline variant of `@urql/exchange-graphcache` I expect to be able to always keep the app usable - not only when the user is offline, but if anything else causes a request to fail I'd always expect the cached data to be displayed.

It works great when the device is offline - then it's possible to browse all the cached data - as one would expect. But if the device is detected as online and receives errors, for example if the server is down and responds with 503, it breaks and urql only exposes the error instead.

The best would be if urql (maybe `offlineExchange` specifically) handles this gracefully out of the box.

Is there a recommended way to handle this?

**Comment by robertherber** [comment]
I'm trying out simplifying the isOfflineError function, like this, it seems to do the trick:
```JavaScript
var isOfflineError = (error) => {
  const isOffline = !!(error && error.networkError);
  return isOffline
}
```

Not sure if I'm missing some scenario here - but maybe this could be considered as default behaviour for the offlineExchange? Potentially with an option to deal "only" with offline errors (i.e. current behaviour)?

**Reply by kitten** [reply]
That's the idea, however, I just realised we forgot to make this function configurable. It's indeed preferable to customise it based on your possible non-GraphQL responses as well. Although, I'm also curious whether we should have better default behaviour here.

The problem is that I'd love to have a blanket solution, however, as always, the edge case in GraphQL are GraphQL-responses (e.g. `{ errors: [...] }`) that are expected. Sometimes it's obviously completely valid and expected for instance to have a `4xx` with a GraphQL response. We could just catch non-data responses with `5xx` status codes but that's obviously also not bullet proof.

So, I think we may want to do the following for Graphcache:
- Extend `storage` with `isOfflineError`
- Provide the current `isOfflineError` as a "base function"
- Provide a default `storage.isOfflineError` that uses the base function and also checks for no-data + 5xx
